src/main/scala/org/broadinstitute/dsde/firecloud/EntityClient.scala

```diff
@@ -111,5 +119,171 @@ class EntityClient (requestContext: RequestContext) extends Actor {
+    ModelSchema.getRequiredAttributes(entityType) match {
+      case Failure(regret) => requestContext.complete(HttpResponse(BadRequest, regret.getMessage))
+      case Success(requiredAttributes) => {
+        if( !requiredAttributes.keySet.subsetOf(headers.toSet) ) {
+          requestContext.complete( HttpResponse(BadRequest, "TSV is missing required attributes: " + (requiredAttributes.keySet -- headers).mkString(", ")) )
+        } else {
+          op(requiredAttributes)
+        }
+      }
+    }
+  }
+
+  def batchCallToRawls( workspaceNamespace: String, workspaceName: String, calls: Seq[EntityUpdateDefinition] ) = {
+    log.info("TSV upload request received")
+    val pipeline: HttpRequest => Future[HttpResponse] =
+      addHeader(Cookie(requestContext.request.cookies)) ~> sendReceive
+    val responseFuture: Future[HttpResponse] = pipeline {
+      Post(FireCloudConfig.Workspace.entityPathFromWorkspace(workspaceNamespace, workspaceName)+"/batchUpsert",
+            HttpEntity(MediaTypes.`application/json`,calls.toJson.toString))
+    }
+
+    responseFuture onComplete {
+      case Success(response) =>
+        response.status match {
+          case NoContent =>
+            log.debug("OK response")
+            requestContext.complete(OK)
+          case _ =>
+            // Bubble up all other unmarshallable responses
+            log.warning("Unanticipated response: " + response.status.defaultMessage)
+            requestContext.complete(response)
+        }
+      case Failure(error) =>
+        // Failure accessing service
+        log.error(error, "Service API call failed")
+        requestContext.failWith(
+          new RequestProcessingException(StatusCodes.InternalServerError, error.getMessage))
+    }
+  }
+
+  val upsertAttrOperation = "op" -> AttributeString("AddUpdateAttribute")
+  val removeAttrOperation = "op" -> AttributeString("RemoveAttribute")
+  val addListMemberOperator = "op" -> AttributeString("AddListMember")
+
+  //colInfo is a list of (headerName, refType), where refType is the type of the entity if the headerName is an AttributeRef
+  //e.g. on TCGA Pairs, there's a header called case_sample_id where the refType would be Sample
+  def setAttributesOnEntity(entityType: String, row: Seq[String], colInfo: Seq[(String,Option[String])]) = {
+    //Iterate over the attribute names and their values
+    //I (hussein) think the refTypeOpt.isDefined is to ensure that if required attributes are left empty, the empty
+    //string gets passed to Rawls, which should error as they're required?
+    val ops = for { (value,(attributeName,refTypeOpt)) <- row.tail zip colInfo if refTypeOpt.isDefined || !value.isEmpty } yield {
+      val nameEntry = "attributeName" -> AttributeString(attributeName)
+      def valEntry( attr: Attribute ) = "addUpdateAttribute" -> attr
+      refTypeOpt match {
+        case Some(refType) => Map(upsertAttrOperation,nameEntry,valEntry(AttributeReference(refType,value)))
```

Comment:
Does __DELETE__ need to be handled here too? How does one remove a member from a list?